finagle-base-http/src/main/scala/com/twitter/finagle/http/MapBackedHeaderMapBase.scala

```diff
@@ -0,0 +1,102 @@
+package com.twitter.finagle.http
+
+import com.twitter.finagle.http.Rfc7230HeaderValidation.{
+  ObsFoldDetected,
+  ValidationFailure,
+  ValidationSuccess
+}
+import com.twitter.logging.Logger
+import scala.annotation.tailrec
+import scala.collection.mutable
+
+/**
+ * Mutable, thread-safe [[HeaderMap]] implementation, backed by
+ * a mutable [[Map[String, Header]]], where the map key
+ * is forced to lower case
+ */
+private[http] abstract class MapBackedHeaderMapBase extends HeaderMap {
+  //implementation methods should be final, but that 
+  //creates a VerifyError: https://github.com/scala/bug/issues/11749
+  import HeaderMap._
+
+  // In general, HashSet/HashTables that are not thread safe are not
+  // durable to concurrent modification and can result in infinite loops.
+  // As such, we synchronize on the underlying `Headers` when performing
+  // accesses to avoid this. In the common case of no concurrent access,
+  // this should be cheap.
+  private[this] val underlying: mutable.Map[String, Header] = mutable.Map.empty
+  
+  //private[this] val underlying: List[Header] = Nil
+  final override def foreach[U](f: ((String, String)) => U): Unit = for {
+    (_, h) <- underlying
+    kv <- h.keyValuePairs
+  } f(kv)
+
+  // ---- HeaderMap -----
+
+  final def getAll(key: String): Seq[String] = underlying.synchronized {
+    underlying.get(key.toLowerCase).toSeq.flatMap(h => h.values)
+  }
+
+  // Validates key and value.
+  final def add(key: String, value: String): HeaderMap = {
+    validateName(key)
+    addUnsafe(key, foldReplacingValidateValue(key, value))
+  }
+
+  // Does not validate key and value.
+  final def addUnsafe(key: String, value: String): HeaderMap = underlying.synchronized {
+    val lower = key.toLowerCase
+    val header = new Header(key, value)
+    underlying.get(lower) match {
+      case Some(h) => h.add(header)
+      case None => underlying.+=((lower, header))
+    }
+    this
+  }
+
+  // Validates key and value.
+  final def set(key: String, value: String): HeaderMap = {
+    validateName(key)
+    setUnsafe(key, foldReplacingValidateValue(key, value))
+  }
+
+  // Does not validate key and value.
+  final def setUnsafe(key: String, value: String): HeaderMap = underlying.synchronized {
+    underlying.+=((key.toLowerCase, new Header(key, value)))
+    this
+  }
+
+  // ---- Map/MapLike -----
+
+  /*final*/ def get(key: String): Option[String] = underlying.synchronized {

```

Comment:
why is this commented out?

Reply:
There was a comment about that. It got lost I think, I'll bring it back. https://github.com/scala/bug/issues/11749